ext/uri/php_uriparser.c

```diff
@@ -0,0 +1,525 @@
+	} else if (uriparser_uri->absolutePath) {
+		ZVAL_CHAR(retval, '/');
+	} else {
+		ZVAL_EMPTY_STRING(retval);
+	}
+
+	return SUCCESS;
+}
+
+static zend_result uriparser_read_query(const uri_internal_t *internal_uri, uri_component_read_mode_t read_mode, zval *retval)
+{
+	uriparser_uris_t *uriparser_uris = (uriparser_uris_t *) internal_uri->uri;
+	UriUriA *uriparser_uri;
+	URIPARSER_READ_URI(uriparser_uri, uriparser_uris, read_mode);
+
+	if (uriparser_uri->query.first != NULL && uriparser_uri->query.afterLast != NULL) {
+		ZVAL_STRINGL(retval, uriparser_uri->query.first, uriparser_uri->query.afterLast - uriparser_uri->query.first);
+	} else {
+		ZVAL_NULL(retval);
+	}
+
+	return SUCCESS;
+}
+
+static zend_result uriparser_read_fragment(const uri_internal_t *internal_uri, uri_component_read_mode_t read_mode, zval *retval)
+{
+	uriparser_uris_t *uriparser_uris = (uriparser_uris_t *) internal_uri->uri;
+	UriUriA *uriparser_uri;
+	URIPARSER_READ_URI(uriparser_uri, uriparser_uris, read_mode);
+
+	if (uriparser_uri->fragment.first != NULL && uriparser_uri->fragment.afterLast != NULL) {
+		ZVAL_STRINGL(retval, uriparser_uri->fragment.first, uriparser_uri->fragment.afterLast - uriparser_uri->fragment.first);
+	} else {
+		ZVAL_NULL(retval);
+	}
+
+	return SUCCESS;
+}
+
+static void *uriparser_malloc(UriMemoryManager *memory_manager, size_t size)
+{
+	return emalloc(size);
+}
+
+static void *uriparser_calloc(UriMemoryManager *memory_manager, size_t nmemb, size_t size)
+{
+	return ecalloc(nmemb, size);
+}
+
+static void *uriparser_realloc(UriMemoryManager *memory_manager, void *ptr, size_t size)
+{
+	return realloc(ptr, size);
+}
+
+static void *uriparser_reallocarray(UriMemoryManager *memory_manager, void *ptr, size_t nmemb, size_t size)
+{
+	const size_t total_size = nmemb * size;
+
+	/* check for unsigned overflow */
+	if ((nmemb != 0) && (total_size / nmemb != size)) {
+		errno = ENOMEM;
+		return NULL;
+	}
+
+	return erealloc(ptr, total_size);
+}
+
+static void uriparser_free(UriMemoryManager *memory_manager, void *ptr)
+{
+	efree(ptr);
+}
+
+zend_result uriparser_request_init(void)
+{
+	uriparser_memory_manager.calloc = uriparser_calloc;
+	uriparser_memory_manager.malloc = uriparser_malloc;
+	uriparser_memory_manager.realloc = uriparser_realloc;
+	uriparser_memory_manager.reallocarray = uriparser_reallocarray;
+	uriparser_memory_manager.free = uriparser_free;
+
+	zend_hash_init(&uriparser_property_handlers, 8, NULL, NULL, true);
+
+	return SUCCESS;
+}
+
+static uriparser_uris_t *uriparser_create_uris(
+	UriUriA *uri, zend_string *uri_str,
+	UriUriA *normalized_uri, zend_string *normalized_uri_str
+) {
+	uriparser_uris_t *uriparser_uris = emalloc(sizeof(uriparser_uris_t));
+
+	uriparser_uris->uri = uri;
+	uriparser_uris->uri_str = uri_str;
+	uriparser_uris->normalized_uri = normalized_uri;
+	uriparser_uris->normalized_uri_str = normalized_uri_str;
+
+	return uriparser_uris;
+}
+
+static void throw_invalid_uri_exception(void)
```

Comment:
Hm don't we have something like this already?

Reply:
In this case the implementation can be simplified to:

```
zend_throw_exception(uri_invalid_uri_exception_ce, "URI parsing failed", 0);
```

Reply:
In fact the implementation for WhatWg can also be simplified to:

```
	zend_object *exception = zend_throw_exception(uri_whatwg_invalid_url_exception_ce, "URL parsing failed", 0);
	zend_update_property(exception->ce, exception, ZEND_STRL("errors"), errors);
```

I can send a PR if desired.

Reply:
#18855